### PR TITLE
Reuse unshifted evaluation when computing parameter shift derivatives of variances

### LINKED ISSUE #3178: Parameter-shift rule of variance does not recycle unshifted evaluation
### Feature details

Save repeated evaluations at unshifted parameters when using `param_shift` together with gradient recipes that involve unshifted evaluations. Consider the setup
```python
import pennylane as qml
import numpy as np

dev = qml.device("default.qubit", wires=2)
x = [0.543, -0.654]

ops_with_custom_recipe = [0]

with qml.tape.QuantumTape() as tape:
    qml.RX(x[0], wires=[0])
    qml.RX(x[1], wires=[0])
    qml.var(qml.PauliZ(0))

gradient_recipes = tuple([[[-1e-7, 1, 0], [1e-7, 1, 0], [-1e5, 1, -5e-6], [1e5, 1, 5e-6]], None])
tapes, fn = qml.gradients.param_shift(tape, gradient_recipes=gradient_recipes)
```
Then we get:
```pycon
>>> for tape in tapes:
>>>     print(tape.get_parameters())
[0.543, -0.654]
[0.543, -0.654]
[0.542995, -0.654]
[0.5430050000000001, -0.654]
[0.543, 0.9167963267948965]
[0.543, -2.2247963267948965]
```
i.e. the unshifted tape is included twice, once for the unshifted evaluation in the product rule of the variance, and once for the shift rule that uses an unshifted term.

### Implementation

This could be saved by passing the information that the unshifted evaluation already is available between `var_param_shift` and the calls to `expval_param_shift`.

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

This issue occurs with and without the new return type system.

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
When computing the derivative of a variance using a gradient recipe with unshifted terms, the unshifted tape is produced and evaluated twice.
See #3178 

**Description of the Change:**
The post-processing function of the expectation value parameter-shift is given an attribute that indicates whether it expects the first result in the results list passed to it to be from the unshifted tape. If this is the case, the variance parameter-shift logic does not add the unshifted expectation value tape $\langle H\rangle$ that is required in the chain rule
$\frac{\partial}{\partial \theta} Var[H] = \frac{\partial}{\partial \theta} \langle H^2\rangle - 2 \langle H\rangle \frac{\partial}{\partial \theta} \langle H\rangle$

**Benefits:**
Saves quantum resources.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
closes #3178

This patch is for both return type systems.

### CODE REVIEW

File: pennylane/gradients/parameter_shift.py [hunk 1]
```diff
@@ -893,7 +897,7 @@ def processing_fn(results):
         shot_vector = isinstance(shots, Sequence)
 
         # analytic derivative of <A>
-        pdA = pdA_fn(results[1:tape_boundary])
+        pdA = pdA_fn(results[int(not pdA_fn.first_result_unshifted) : tape_boundary])
```
Comment (rmoyard) [inline pennylane/gradients/parameter_shift.py:900]: Here before it was always [1:tape_boundary], now if unshifted we have [1:tape_boundary], and if not unshifted [0:tape_boundary]. So why you need to consider more tape for the not unshifted case?

File: pennylane/gradients/parameter_shift.py [hunk 2]
```diff
@@ -980,11 +982,12 @@ def _var_param_shift_tuple(
     pdA_tapes, pdA_fn = expval_param_shift(
         expval_tape, argnum, shifts, gradient_recipes, f0, broadcast, shots
     )
+    gradient_tapes = [] if pdA_fn.first_result_unshifted else [expval_tape]
```
Comment (rmoyard) [inline pennylane/gradients/parameter_shift.py:985]: Great to have less tapes if we have unshifted tapes :+1:

**Review by rmoyard — COMMENTED** [review]
Great to save some quantum evaluations :100: I have one question after that happy to approve.

File: pennylane/gradients/parameter_shift.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/gradients/parameter_shift.py:900]: Ah, so the logic is as follows: If the first tape created by the expectation value parameter-shift rule is *not* the unshifted tape (`not pdA_fn.first_result_unshifted = True`), the variance computation requires an unshifted tape and has to add it. In that case, the expectation value results (passed to `pdA_fn`) start with the second entry, the first one is the unshifted one.
If the first tape created by the expectation value parameter-shift rule *is* the unshifted tape (`not pdA_fn.first_result_unshifted = False`), the variance computation does not need to add a tape, so the results passed to `pdA_fn` just start with the very first result :)

**Review by rmoyard — APPROVED** [review]
Thanks nice job :100:

**Pull request merged**